File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/api/PolarH10OfflineExerciseApi.swift
//  Copyright © 2023 Polar. All rights reserved.

import Foundation
import RxSwift

///  Recoding intervals for H10 recording start
public enum RecordingInterval: Int {
    case interval_1s = 1
    case interval_5s = 5
}

/// Sample types for H10 recording start
public enum SampleType: Int {
    /// recording type to use is hr in BPM
    case hr
    /// recording type to use is rr interval
    case rr
}

/// H10 Exercise recording API.
///
/// H10 Exercise recording makes it possible to record Hr or Rr data to H10 device memory.
/// With H10 Exercise recording the H10 and phone don't need to be connected all the time, as H10 exercise recording
/// continues in Polar device even the BLE disconnects.
///
/// Requires features `PolarBleSdkFeature.feature_polar_h10_exercise_recording`
///
/// Note, API is working only with Polar H10 device
///
public protocol PolarH10OfflineExerciseApi {
    /// Request start recording. Supported only by Polar H10. Requires `polarFileTransfer` feature.
    ///
    /// - Parameters:
    ///   - identifier: Polar device id or UUID
    ///   - exerciseId: unique identifier for for exercise entry length from 1-64 bytes
    ///   - interval: recording interval to be used. Has no effect if `sampleType` is `SampleType.rr`
    ///   - sampleType: sample type to be used.
    /// - Returns: Completable stream
    ///   - success: recording started
    ///   - onError: see `PolarErrors` for possible errors invoked
    func startRecording(_ identifier: String, exerciseId: String, interval: RecordingInterval, sampleType: SampleType) -> Completable
    
    /// Request stop for current recording. Supported only by Polar H10. Requires `polarFileTransfer` feature.
    ///
    /// - Parameters:
    ///   - identifier: Polar device id or UUID
    /// - Returns: Completable stream
    ///   - success: recording stopped
    ///   - onError: see `PolarErrors` for possible errors invoked
    func stopRecording(_ identifier: String) -> Completable
    
    /// Request current recording status. Supported only by Polar H10. Requires `polarFileTransfer` feature.
    ///
    /// - Parameters:
    ///   - identifier: Polar device id
    /// - Returns: Single stream
    ///   - success: see `PolarRecordingStatus`
    ///   - onError: see `PolarErrors` for possible errors invoked
    func requestRecordingStatus(_ identifier: String) -> Single<PolarRecordingStatus>
    
    /// Api for fetching stored exercises list from Polar H10 device. Requires `polarFileTransfer` feature. This API is working for Polar OH1 and Polar Verity Sense devices too, however in those devices recording of exercise requires that sensor is registered to Polar Flow account.
    ///
    /// - Parameters:
    ///   - identifier: Polar device id or device address
    /// - Returns: Observable stream
    ///   - onNext: see `PolarExerciseEntry`
    ///   - onError: see `PolarErrors` for possible errors invoked
    func fetchStoredExerciseList(_ identifier: String) -> Observable<PolarExerciseEntry>
    
    /// Api for fetching a single exercise from Polar H10 device. Requires `polarFileTransfer` feature. This API is working for Polar OH1 and Polar Verity Sense devices too, however in those devices recording of exercise requires that sensor is registered to Polar Flow account.
    ///
    /// - Parameters:
    ///   - identifier: Polar device id or device address
    ///   - entry: single exercise entry to be fetched
    /// - Returns: Single stream
    ///   - success: invoked after exercise data has been fetched from the device. see `PolarExerciseEntry`
    ///   - onError: see `PolarErrors` for possible errors invoked
    func fetchExercise(_ identifier: String, entry: PolarExerciseEntry) -> Single<PolarExerciseData>
    
    /// Api for removing single exercise from Polar H10 device. Requires `polarFileTransfer` feature. This API is working for Polar OH1 and Polar Verity Sense devices too, however in those devices recording of exercise requires that sensor is registered to Polar Flow account.
    ///
    /// - Parameters:
    ///   - identifier: Polar device id or device address
    ///   - entry: single exercise entry to be removed
    /// - Returns: Completable stream
    ///   - complete: entry successfully removed
    ///   - onError: see `PolarErrors` for possible errors invoked
    func removeExercise(_ identifier: String, entry: PolarExerciseEntry) ->Completable
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/api/PolarH10OfflineExerciseApi.kt
// Copyright © 2023 Polar Electro Oy. All rights reserved.
package com.polar.sdk.api

import androidx.annotation.Size
import androidx.core.util.Pair
import com.polar.sdk.api.model.PolarExerciseData
import com.polar.sdk.api.model.PolarExerciseEntry
import kotlinx.coroutines.flow.Flow

/**
 * H10 Exercise recording API.
 *
 * H10 Exercise recording makes it possible to record Hr or Rr data to H10 device memory.
 * With H10 Exercise recording the H10 and phone don't need to be connected all the time, as H10 exercise recording
 * continues in Polar device even the BLE disconnects.
 *
 * Requires features [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_H10_EXERCISE_RECORDING]
 *
 * Note, API is working only with Polar H10 device
 */
interface PolarH10OfflineExerciseApi {

    /**
     * Recoding intervals for H10 recording start
     */
    enum class RecordingInterval(val value: Int) {
        INTERVAL_1S(1), /*!< 1 second interval */
        INTERVAL_5S(5); /*!< 5 second interval */
    }

    /**
     * Sample types for H10 recording start
     */
    enum class SampleType {
        HR, /*!< HeartRate in BPM */
        RR, /*!< RR interval in milliseconds */
    }

    /**
     * Request start recording. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_H10_EXERCISE_RECORDING]
     *
     * @param identifier polar device id or bt address
     * @param exerciseId unique id for exercise entry
     * @param interval   recording interval to be used, parameter has no effect if the `type` parameter is SampleType.RR
     * @param type       sample type to be used
     * @return Success or error
     */
    suspend fun startRecording(
        identifier: String,
        @Size(min = 1, max = 64) exerciseId: String,
        interval: RecordingInterval?,
        type: SampleType
    )

    /**
     * Request to stop recording. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_H10_EXERCISE_RECORDING]
     *
     * @param identifier polar device id or bt address
     * @return Success or error
     */
    suspend fun stopRecording(identifier: String)

    /**
     * Request current recording status. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_H10_EXERCISE_RECORDING]
     *
     * @param identifier polar device id or bt address
     * @return Pair first recording status, second entryId if available
     */
    suspend fun requestRecordingStatus(identifier: String): Pair<Boolean, String>

    /**
     * List exercises stored in the device Polar H10 device. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_H10_EXERCISE_RECORDING]
     *
     * @param identifier Polar device id found printed on the sensor/device or bt address
     * @return [Flow] stream of [PolarExerciseEntry] entries
     */
    fun listExercises(identifier: String): Flow<PolarExerciseEntry>

    /**
     * Api for fetching a single exercise from Polar H10 device. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_H10_EXERCISE_RECORDING]
     *
     * @param identifier Polar device id found printed on the sensor/device or bt address
     * @param entry      [PolarExerciseEntry] object
     * @return [PolarExerciseData]
     */
    suspend fun fetchExercise(identifier: String, entry: PolarExerciseEntry): PolarExerciseData

    /**
     * Api for removing single exercise from Polar H10 device. Requires feature [PolarBleApi.PolarBleSdkFeature.FEATURE_POLAR_H10_EXERCISE_RECORDING]
     *
     * @param identifier Polar device id found printed on the sensor/device or bt address
     * @param entry      entry to be removed
     */
    suspend fun removeExercise(identifier: String, entry: PolarExerciseEntry)
}